pkg/network/encoding/http2.go

```diff
@@ -6,152 +6,90 @@
+		},
 	}
-	return nil, 0, nil
-}
 
-func (a *http2AggregationWrapper) ValueFor(c network.ConnectionStats) *model.HTTP2Aggregations {
-	if a == nil {
-		return nil
+	http2StatusCodeMap = sync.Pool{
+		New: func() any {
+			m := make(map[int32]*model.HTTPStats_Data)
+			return &m
+		},
 	}
+)
 
-	if a.sport == 0 && a.dport == 0 {
-		// This is the first time a ConnectionStats claim this aggregation. In
-		// this case we return the value and save the source and destination
-		// ports
-		a.sport = c.SPort
-		a.dport = c.DPort
-		return a.HTTP2Aggregations
-	}
+type http2Encoder struct {
+	byConnection *USMConnectionIndex[http.Key, *http.RequestStats]
 
-	if c.SPort == a.dport && c.DPort == a.sport {
-		// We have have a collision with another `ConnectionStats`, but this is a
-		// legit scenario where we're dealing with the opposite ends of the
-		// same connection, which means both server and client are in the same host.
-		// In this particular case it is correct to have both connections
-		// (client:server and server:client) referencing the same HTTP2 data.
-		return a.HTTP2Aggregations
-	}
+	// cached object
+	aggregations *model.HTTP2Aggregations
 
-	// Return nil otherwise. This is to prevent multiple `ConnectionStats` with
-	// exactly the same source and destination addresses but different PIDs to
-	// "bind" to the same HTTP2Aggregations object, which would result in a
-	// overcount problem. (Note that this is due to the fact that
-	// `types.ConnectionKey` doesn't have a PID field.) This happens mostly in the
-	// context of pre-fork web servers, where multiple worker processes share the
-	// same socket
-	return nil
+	// list of *pointers* to maps so they can be returned to the pool
+	toRelease []*map[int32]*model.HTTPStats_Data
 }
 
 func newHTTP2Encoder(payload *network.Connections) *http2Encoder {
 	if len(payload.HTTP2) == 0 {
 		return nil
 	}
 
-	encoder := &http2Encoder{
-		aggregations:   make(map[types.ConnectionKey]*http2AggregationWrapper, len(payload.Conns)),
-		staticTags:     make(map[types.ConnectionKey]uint64, len(payload.Conns)),
-		dynamicTagsSet: make(map[types.ConnectionKey]map[string]struct{}, len(payload.Conns)),
+	return &http2Encoder{
+		byConnection: GroupByConnection("http2", payload.HTTP2, func(key http.Key) types.ConnectionKey {
+			return key.ConnectionKey
+		}),
+		aggregations: new(model.HTTP2Aggregations),
 	}
+}
 
-	// pre-populate aggregation map with keys for all existent connections
-	// this allows us to skip encoding orphan HTTP2 objects that can't be matched to a connection
-	for _, conn := range payload.Conns {
-		for _, key := range network.ConnectionKeysFromConnectionStats(conn) {
-			if log.ShouldLog(seelog.TraceLvl) {
-				log.Tracef("Payload has a connection %v and was converted to http2 key %v", conn, key)
-			}
-			encoder.aggregations[key] = nil
-		}
+func (e *http2Encoder) GetHTTP2AggregationsAndTags(c network.ConnectionStats) ([]byte, uint64, map[string]struct{}) {
```

Comment:
Question about this function that may be out of context for this PR:
why does the function GetHTTP2AggregationsAndTags is returning ([]byte, uint64, map[string]struct{}) seems like we do not accept the second and third args in  the code. (expect from UTs)

Reply:
this is coming from the original code, the refactor does not change that
the issue is with merging the tags with the http tags.
currently it is not very relevant as we don't have tags on http2 connetion (the http tags are mainly around tls library),  in the future we should merge the tags

Reply:
so can you create a Jira task for that? in order for us to remember to return to this function and refactor it

Reply:
done